enhancements/installer/openstack-baremetal-compute.md

```diff
@@ -0,0 +1,156 @@
+be used as OpenShift Compute Nodes, when they are made available as Nova
+flavors.
+
+## Motivation
+
+In heavy-loaded clusters, or when the workload is better handled with specially
+purposed hardware, the functional requirement of controlling where the work is
+computed might outweight the added value of virtualisation.
+
+In OpenStack clusters where Nova transparently provisions baremetal machines,
+this enhancement will ensure that OpenShift installations can use them as
+Compute Nodes.
+
+### Goals
+
+* Enable running OpenShift on a mix of virtual and baremetal machines.
+* Baremetal machines can be added and removed from a running OpenShift cluster.
+
+### Non-Goals
+
+**Separate subnets.**
+
+An OpenStack setting where baremetal machines can't be attached to a virtual
+subnet is not supported in this enhancement.
+
+**External connectivity via floating IPs.**
+
+The OpenStack cluster might make the BM nodes accessible, but exposing them via
+floating IPs might represent an increase in complexity and is currently not a
+goal.
+
+## Proposal
+
+This enhancement is less about adding a new logical piece, than it is about
+uncovering unknowns in a configuration that might work out-of-the-box in
+selected environments.
+
+Therefore, the implementation consists of:
+* identifying a reference architecture
+* anticipating probable issues
+* testing and fixing errors
+
+### Characteristics of the reference architecture
+
+* The choice between deploying a VM or BM node is done by selecting an appropriate Nova flavour
+  * which can be done both in install config and passed to Cluster API provider OpenStack
+  * for the time being, the interactive prompt does ask the flavor for the Compute Nodes specifically; the choice must be made editing `install-config.yaml`
+* The deployment request makes the same API call to Nova as with a VM install (there are no explicit calls to Ironic or anything in the Installer required)
+  * so creating baremetal nodes should work transparently from the installer as well as Cluster API provider OpenStack
+* both VMs and BMs consume the same image from glance
+
+### Probable issues
+
+* The Ignition boot system consisting on a second payload downloaded at runtime
+  requires the baremetal machine to have access to Swift early in the boot
+  process.
+
+* When attaching and detaching, the expected timings must be adjusted to
+  reflect the different boot and shutdown latencies of physical machines
+
+* When attaching, the Certificate Signing Request timeout must be adjusted to
+  reflect the different boot and shutdown latencies of physical machines
+
+### User Stories
+
+#### Increased performance
+
+User wants to run the Compute Nodes on baremetal machines for increased
+performance. The Control Plane sits on regular Nova instances to leverage the
+flexibility of virtual machines.
+
+### Risks and Mitigations
+
+TBD
+
+## Design Details
+
+### Test Plan
+
+The enhancement will be tested using virtual baremetal machines.
```

Comment:
@tomassedovic I don't know about OVB, but for Metal3 we've largely been using virtualbmc and sushy-tools... and leaning more towards sushy tools as we have a need to support redfish.  Ultimately I guess it all comes down to how are you going to prepare your test environment, and is there _really_ anything special that needs to be implemented in order to test functionality? There shouldn't be differences, but it also entirely depends on how your going about asking for instances from the IaaS.

@mandre Most of the team and community is out until January. I can't commit anyone's time or spoons, but I suspect if a short call with specific questions were to be scheduled for context sharing, that might be able to help without taxing the internal team too much.

Reply:
I've tried to get OVB working with OpenShift, but it's problematic because the bootstrap VM for baremetal is hard-coded as a libvirt VM, which means nested virt. However, if you're bootstrapping in OpenStack then you could probably use it.

Currently for baremetal we're just using libvirt vms configured by https://github.com/metal3-io/metal3-dev-env/ but that has all of the drawbacks that led to OVB in the first place.

Reply:
Ah right, virtualbmc and sushy! I did seem to remember the Ironic folks using something other than OVB, but couldn't remember what. Thanks so much @juliakreger!